File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool check(vector<int>& weights,int days,int m)
    {
         int n=weights.size();
         int count=1;
         int sum=0;
         for(int i=0;i<n;i++)
         {
            sum+=weights[i];
            if(sum>m)
            {
                count++;
                sum=weights[i];
            }
         }
         return count<=days;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int res=INT_MAX;

        int l=*max_element(weights.begin(),weights.end());
        int h=accumulate(weights.begin(),weights.end(),0);
        while(l<=h)
        {
            int m=l+(h-l)/2;
            if(check(weights,days,m))
            {
                res=m;
                h=m-1;
            }
            else
            {
                l=m+1;
            }
        }
        return res;
    }
};

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    public boolean check(int []weights,int mid,int days)
    {
        int sum=0,cnt=1;
        int n=weights.length;
        for(int i=0;i<n;i++)
        {
            if(sum+weights[i]<=mid)
            {
                sum+=weights[i];
            }else{
                cnt++;
                sum=weights[i];
            }
        }
        return cnt<=days;
    }
    public int shipWithinDays(int[] weights, int days) {
        int n=weights.length;
        int ans=0;
        int l=0,h=0;

        for(int num:weights)
        {
            l=Math.max(l,num);
            h+=num;
        }

        while(l<=h)
        {
            int mid=l+(h-l)/2;

            if(check(weights,mid,days))
            {
                ans=mid;
                h=mid-1;
            }else
            {
                l=mid+1;
            }
        }
        return ans;
    }
}